flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/TaskExecutionStateWrapper.java

```diff
@@ -0,0 +1,102 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.executiongraph;
+
+import org.apache.flink.api.common.JobID;
+import org.apache.flink.runtime.accumulators.AccumulatorSnapshot;
+import org.apache.flink.runtime.execution.ExecutionState;
+import org.apache.flink.runtime.taskmanager.TaskExecutionState;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+import static org.apache.flink.util.Preconditions.checkState;
+
+/**
+ * Wraps {@link TaskExecutionState}, along with actions to take
+ * if it is FAILED state.
+ */
+public class TaskExecutionStateWrapper {
+
+	private final TaskExecutionState taskExecutionState;
+
+	/**
+	 * Indicating whether to send a PRC call to remove task from TaskManager.
+	 * True if the failure is fired by JobManager and the execution is already
+	 * deployed. Otherwise it should be false.
+	 */
+	private final boolean cancelTask;
+
+	private final boolean releasePartitions;
+
+	private TaskExecutionStateWrapper(
+			final TaskExecutionState taskExecutionState,
+			final boolean cancelTask,
+			final boolean releasePartitions) {
+
+		this.taskExecutionState = checkNotNull(taskExecutionState);
+		this.cancelTask = cancelTask;
+		this.releasePartitions = releasePartitions;
+	}
+
+	public Throwable getError(ClassLoader userCodeClassloader) {
+		return taskExecutionState.getError(userCodeClassloader);
+	}
+
+	public ExecutionAttemptID getID() {
+		return taskExecutionState.getID();
+	}
+
+	public ExecutionState getExecutionState() {
+		return taskExecutionState.getExecutionState();
+	}
+
+	public JobID getJobID() {
+		return taskExecutionState.getJobID();
+	}
+
+	public AccumulatorSnapshot getAccumulators() {
+		return taskExecutionState.getAccumulators();
+	}
+
+	public IOMetrics getIOMetrics() {
+		return taskExecutionState.getIOMetrics();
+	}
+
+	public boolean getCancelTask() {
+		checkState(taskExecutionState.getExecutionState() == ExecutionState.FAILED);
+		return cancelTask;
+	}
+
+	public boolean getReleasePartitions() {
+		checkState(taskExecutionState.getExecutionState() == ExecutionState.FAILED);
+		return releasePartitions;
+	}
+
+	public static TaskExecutionStateWrapper createFrom(TaskExecutionState taskExecutionState) {
+		return new TaskExecutionStateWrapper(taskExecutionState, false, false);
+	}
+
+	public static TaskExecutionStateWrapper createFromFailedState(
+			final TaskExecutionState taskExecutionState,
+			final boolean cancelTask,
+			final boolean releasePartitions) {
+
+		checkState(taskExecutionState.getExecutionState() == ExecutionState.FAILED);
```

Comment:
it is not intuitive that this method checks for a failed state, but the plain `createFrom()` does not assert the opposite (i.e., that the task did not fail).

Reply:
I will remove the check and the factory methods. See comment https://github.com/apache/flink/pull/13766#discussion_r514952083.

Reply:
Just a note that `createFrom()` will also used for FAILED state which is reported from TaskManager because the flags will be false.